core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java

```diff
@@ -1887,7 +1886,7 @@ public PhysicalOperation visitGroupId(GroupIdNode node, LocalExecutionPlanContex
 
             int outputChannel = 0;
 
-            for (Symbol output : node.getGroupingSets().stream().flatMap(Collection::stream).collect(Collectors.toSet())) {
```

Comment:
would it be enough to say `toImmutableSet` here?

Reply:
At this particular usage site it would be, but the same logic of producing distinct grouping set symbols is used in a few other places inside of `GroupIdNode` so I think it makes sense to just use it as a public method.

Reply:
i think it would be nice to separate "fix stability" from "reduce code duplication". maybe two commits?

Reply:
Unfortunately all of these changes are necessary to avoid the usage of unstable-ordering `Set` implementations, so breaking the commit into two would just introduce temporary changes that would immediately be deleted by the code de-duplication refactor- so I think it's simpler to leave it as a single commit.